keps/sig-api-machinery/2211-omit-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,372 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+ManagedFields is used to keep track of who has changed a field of an object, and Server-side Apply relies on such information.
+While the managedFields is beneficial to resolving conflicts on updates, it is a bit verbose to many users and controllers
+because this field could be quite large and generally users and controllers are not interested in such information. 
+In addition, the bandwidth usage and size of internal caches could be significantly affected due to the additional managedFields.
+
+This KEP is proposing omitting the managedFields on the server-side when clients request objects to make the response less verbose
+and save the amount of data that needs to be transferred.
+
+## Motivation
+
+Many users have complained in [kubernetes#90066][90066] that they got impacted by the addition of managedFields
+(and the issue itself has received more than 100 thumb-up reactions), and we want to improve the user experience.
+
+[90066]: https://github.com/kubernetes/kubernetes/issues/90066
+
+At the same time, it is not ideal to only drop the managedFields on the client-side, because that requires the same
+bandwidth usage compared to simply returning the original object, and every client
+(such as kubectl or CLI tools written by users using client-go) needs to drop the managedFields by themselves.
+
+With this KEP, the API server could return objects without managedFields on demand, so that least client-side changes is needed
+to make the response less verbose. 
+
+### Goals
+
+* Omit managedFields from responses, when requested by clients.
+
+### Non-Goals
+
+* Turn off managedFields. Server-side Apply relies on managedFields, we could not simply disable it.
+* Clear manageFields only on the client-side.
+
+## Proposal
+
+### Server-side changes
+
+* Add three content types:
+  * `application/json+managed.omit`
+  * `application/yaml+managed.omit`
+  * `application/vnd.kubernetes.protobuf+managed.omit`
+* Add a new serializer, say `UnmanagedSerializer`, which wraps existing serializers and omits the managedFields before encoding objects.
+* Make the API server to use `UnmanagedSerializer` when clients properly set the `Accept` header and the selected
+  content type after negotiation is one of the above three.
+
+In addition, it is worth noting that we could add similar content types to represent specific format version of managedFields,
+such as `application/json+managed.v1`, `application/yaml+managed.v2` etc, so that users are able to request a certain
+version of managedFields with these headers.
+
+### Client-side changes
+
+Implement a flag in `kubectl` to let users choose to omit the managedFields.
```

Comment:
Did we consider adding a line to kubeconfig files, to make it easier to apply this to controller requests? (i.e., workflow would be change kubeconfig, rebase controller on latest client & recompile.) @deads2k might have thoughts on whether this is reasonable to do to kubeconfig files.

Reply:
One concern is that it makes debugging harder.
> - Please run `kubectl get ... -o yaml`
> - Sure, here's the output: `yaml without managed fields`
> - Why are `managedFields` missing? Ah, each person has a different output based on some "hidden" configuration. 

These are situations typically confusing for everyone.